streamz/tests/test_plugins.py

```diff
@@ -0,0 +1,69 @@
+import pytest
+from streamz import Source, Stream
+
+
+class MockEntryPoint:
+
+    def __init__(self, name, cls, module_name=None):
+        self.name = name
+        self.cls = cls
+        self.module_name = module_name
+
+    def load(self):
+        return self.cls
+
+
+def test_register_plugin_entry_point():
+    class test_stream(Stream):
+        pass
+
+    entry_point = MockEntryPoint("test_node", test_stream)
+    Stream.register_plugin_entry_point(entry_point)
+
+    assert Stream.test_node.__name__ == "stub"
+
+    Stream().test_node()
+
+    assert Stream.test_node.__name__ == "test_stream"
+
+
+def test_register_plugin_entry_point_modifier():
+    class test_source(Source):
+        pass
+
+    def modifier(fn):
+        fn.__name__ = 'modified_name'
+        return staticmethod(fn)
+
+    entry_point = MockEntryPoint("from_test", test_source)
+    Stream.register_plugin_entry_point(entry_point, modifier)
+
+    Stream.from_test()
+
+    assert Stream.from_test.__name__ == "modified_name"
+
+
+def test_register_plugin_entry_point_raises():
+    class invalid_node:
+        pass
+
+    entry_point = MockEntryPoint("test", invalid_node, "test_module.test")
+
+    Stream.register_plugin_entry_point(entry_point)
+
+    with pytest.raises(TypeError):
+        Stream.test()
+
+
+def test_register_plugin_entry_point_already_registered():
+    @Stream.register_api()
+    class test(Stream):
+        pass
+
+    entry_point = MockEntryPoint("test_double", test, "test_module")
+
+    Stream.register_plugin_entry_point(entry_point)
+
+    assert Stream.test_double.__name__ == "stub"
```

Comment:
Actually, I'm not sure whether the method should have been overwritten - presumably `Stream.test_double` already existed before the call to `register_plugin_entry_point`.

In normal usage, the stubs always come first, since they happen at import time, and streamz should always be imported before derived classes.

Reply:
Yes, I guess this test actually doesn't test anything. I wanted to test that running `Stream.register_api` twice on the same class doesn't break anything, in case the implementation of `register_api` changes. We can just check for this in `register_plugin_entry_point`.

Reply:
The test is OK, but not comprehensive. The cases are:
- register_api, then register entrypoint
- register entrypoint then register_api
- either of the to methods called twice

In every case, calling the method should result in the same test class.

We *could* check for whether the name is being overwritten, but I don't think that's essential.

Reply:
Btw, what should we do if a plugin wants to override a built-in method? On the one hand, this would allow people to create useful extensions of built-in functionality (my version of `partition` could be provided as a plugin long before it's merged into core), on the other — it could break things and lead to confusion.

Reply:
Personally I would have the system balk at a built-in override. My hope would be that names are cheap, so adding more text onto a would be `partition` describing how it is difficult would add less burden than issues around which `partition` was used. This could get particularly thorny as envs get updated, code that previously worked and relied on the built-in are now producing errors in subtle ways that could be difficult to find.

Reply:
> In every case, calling the method should result in the same test class.

The class is stored in `Stream.method.__wrapped__`, but I don't see how anything else but the test class can end up there in any of these cases.

Reply:
You mean we should have a `clobber=False` argument for the register methods? 
I wonder in that case, if there is a sane way to tell when an entrypoint and register_api refer to the same thing, which would not be a problem even without clobber.

Reply:
> I wonder in that case, if there is a sane way to tell when an entrypoint and register_api refer to the same thing, which would not be a problem even without clobber.

We can just check that `hasattr(Stream, entry_point.name)`. But would have to happen during plugin load (so that we're not overwriting built-in methods) and so importing streamz with a bad plugin would result in an error.

Reply:
but this will fail when someone adds the entrypoint to a class that already has register_api, no?

Note @CJ-Wright , that the current implementation does allow you to use register_api to overwrite methods at will. Of course this is python, so we can't ever disable someone who wants to do that anyway.

Reply:
That's fair. To some extent this boils down to locality for me. If the author of the code is the user of the env then if they override things then they are in a better place to clean up. When installing 3rd party things that state may not hold as well.

Reply:
> If the author of the code is the user of the env then if they override things then they are in a better place to clean up. When installing 3rd party things that state may not hold as well.

I agree. With `register_api` things are explicit and visible. A plugin is a black box.

@martindurant please look at the last commit to see what I mean